packages/blade/src/components/ActionList/ActionList.tsx

```diff
@@ -1,115 +1,136 @@
-        selectInputRef.current?.focus();
-      }}
-      onMouseDown={() => {
-        setShouldIgnoreBlur(true);
-      }}
-      href={props.href}
-      style={{
-        border: activeIndex === props.index ? '2px solid red' : '',
-        width: '100%',
-      }}
-    >
-      <Text>{props.title}</Text>
-    </Box>
-  );
-};
 
-type ActionListProps = {
-  children: React.ReactNode[];
-};
-const ActionList = ({ children }: ActionListProps): JSX.Element => {
-  const { setOptions, actionListRef, selectionType, dropdownBaseId } = useDropdown();
-  const actionListOptions: {
-    title: string;
-    value: string;
-  }[] = [];
+      const clonedChild = React.cloneElement(child, {
+        // @ts-expect-error: TS doesn't understand the child's props
+        index: currentIndex,
+      });
+      return clonedChild;
+    }
+
+    return child;
+  };
+
+  const childCompIdArray = React.Children.toArray(children).map((child) => getComponentId(child));
+  const lastActionListSectionIndex = childCompIdArray.lastIndexOf(componentIds.ActionListSection);
+
+  const isActionListItemPresentAfterSection = childCompIdArray
+    .slice(lastActionListSectionIndex)
+    .includes(componentIds.ActionListItem);
 
   // Looping through ActionListItems to add index to them and get an options array for moving focus between items
-  const childrenWithId = React.Children.map(children, (child) => {
+  const childrenWithId = React.Children.map(children, (child, index) => {
     if (React.isValidElement(child)) {
-      // @TODO: handle the scenario where ActionListItem is inside ActionListMenu
-      if (child.type === ActionListItem) {
-        actionListOptions.push({
-          title: child.props.title,
-          value: child.props.value,
-        });
-        const currentIndex = actionListOptions.length - 1;
-        const clonedChild = React.cloneElement(child, {
+      if (isValidAllowedChildren(child, componentIds.ActionListSection)) {
+        return React.cloneElement(child, {
           // @ts-expect-error: TS doesn't understand the child's props
-          index: currentIndex,
+          children: React.Children.map(child.props.children, (childInSection) => {
+            if (isValidAllowedChildren(childInSection, componentIds.ActionListItem)) {
+              return getActionListItemChildWithId(childInSection);
+            }
+
+            return childInSection;
+          }),
+          hideDivider: index === lastActionListSectionIndex && !isActionListItemPresentAfterSection,
         });
-        return clonedChild;
+      }
+      if (isValidAllowedChildren(child, componentIds.ActionListItem)) {
+        return getActionListItemChildWithId(child);
       }
     }
-
     return child;
   });
 
   React.useEffect(() => {
     setOptions(actionListOptions);
     // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [setOptions, children]);
+  }, [optionsRecalculateToggle]);
```

Comment:
What does this `optionsRecalculateToggle` do?

Reply:
It is a toggle flag which I change to re-run this useEffect. So if there are any errors in loading options, it just retriggers this function.

Reply:
>  So if there are any errors in loading options, it just retriggers this function.

What kind of errors?

Reply:
yes. This is mostly try catch equivalent. I haven't found any case where this happens.